src/main/java/org/gradle/java/JigsawPlugin.java

```diff
@@ -1,15 +1,111 @@
+/*
+ * Copyright 2009 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package org.gradle.java;
 
+import org.gradle.api.Action;
+import org.gradle.api.GradleException;
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
+import org.gradle.api.Task;
+import org.gradle.api.file.CopySpec;
 import org.gradle.api.logging.Logger;
 import org.gradle.api.logging.Logging;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.SourceSetContainer;
+import org.gradle.api.tasks.compile.JavaCompile;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
 
 public class JigsawPlugin implements Plugin<Project> {
     private static final Logger LOGGER = Logging.getLogger(JigsawPlugin.class);
 
+    private static final String JAVA_PLUGIN = "org.gradle.java";
+
+    private static final String JAVA_LIBRARY_PLUGIN = "org.gradle.java-library";
+
+    private static final String APPLICATION_PLUGIN = "application";
+
     @Override
     public void apply(Project project) {
         LOGGER.debug("Applying JigsawPlugin to " + project.getName());
+
+        project.getExtensions().create("jigsawModule", JigsawModule.class);
+
+        configureJavaTasks(project);
+    }
+
+    private void configureJavaTasks(Project project) {
+        project.afterEvaluate(new Action<Project>() {
+            @Override
+            public void execute(Project project) {
+                if (project.getPlugins().hasPlugin(JAVA_PLUGIN) || project.getPlugins().hasPlugin(JAVA_LIBRARY_PLUGIN)) {
+                    configureCompileJavaTask(project);
+                } else {
+                    throw new GradleException("You must apply the java or java-library plugin as well.");
+                }
+            }
+        });
+    }
+
+    private void configureCompileJavaTask(final Project project) {
+        final JavaCompile compileJava = getCompileJavaTask(project);
+        final SourceSet main = ((SourceSetContainer) project.getProperties().get("sourceSets")).getByName("main");
+        final JigsawModule module = (JigsawModule) project.getExtensions().getByName("jigsawModule");
```

Comment:
When you say "bind the module name to Jigsaw" you're saying you think the extension name shouldn't be `jigsawModule`. I don't really like the idea of changing the name of the extension to `api` because I don't think that's really what we're modeling. Conceptually, this really should be the `project.name` (as long as we are only generating one module per project.) The module doesn't just consist of the exported packages, but actually all the classes defined in the `main` sourceset. Maybe, it should really be a `moduleName` property on sourceset. But, if it should, I feel like that's too big of a change for this plugin to handle. For now though, I agree that `jigsawModule.name` is too specific. I originally used `module.name` but apparently that clashes with an internal property of Project.  I get this message when I try to use it:
`Cannot set readonly property: name for class: org.gradle.api.internal.artifacts.ProjectBackedModule`

Maybe we can use one of these names for the extension point?
`globalModule.name`
`advertisedModule.name`
`exportedModule.name`
`encapsulation.moduleName`
`architecture.moduleName`
`structure.moduleName`
`public.moduleName`

I feel like we're suffering from `TooManyModulesException`. For now, I've called it `architecture.moduleName` because I think it's better than `jigsawModule.name` and we can argue about the right name more in this code review.